halo2_proofs/src/poly/kzg/commitment.rs

```diff
@@ -0,0 +1,329 @@
+//! This module contains an implementation of the polynomial commitment scheme
+//! described in the [Halo][halo] paper.
+//!
+//! [halo]: https://eprint.iacr.org/2019/1021
+
+use crate::arithmetic::{
+    best_fft, best_multiexp, parallelize, CurveAffine, CurveExt, FieldExt, Group,
+};
+use crate::helpers::CurveRead;
+use crate::poly::commitment::{Blind, CommitmentScheme, Params, ParamsProver, ParamsVerifier, MSM};
+use crate::poly::{Coeff, LagrangeCoeff, Polynomial};
+
+use ff::{Field, PrimeField};
+use group::{prime::PrimeCurveAffine, Curve, Group as _};
+use halo2curves::pairing::Engine;
+use std::fmt::Debug;
+use std::marker::PhantomData;
+use std::ops::{Add, AddAssign, Mul, MulAssign};
+
+use std::io;
+
+use super::msm::MSMKZG;
+
+/// These are the public parameters for the polynomial commitment scheme.
+#[derive(Debug, Clone)]
+pub struct ParamsKZG<E: Engine> {
+    pub(crate) k: u32,
+    pub(crate) n: u64,
+    pub(crate) g: Vec<E::G1Affine>,
+    pub(crate) g_lagrange: Vec<E::G1Affine>,
+    pub(crate) g2: E::G2Affine,
+    pub(crate) s_g2: E::G2Affine,
+}
+
+/// Umbrella commitment scheme construction for all KZG variants
+#[derive(Debug)]
+pub struct KZGCommitmentScheme<E: Engine> {
+    _marker: PhantomData<E>,
+}
+
+impl<'params, E: Engine + Debug> CommitmentScheme<'params> for KZGCommitmentScheme<E> {
+    type Scalar = E::Scalar;
+    type Curve = E::G1Affine;
+    type MSM = MSMKZG<E>;
+
+    type ParamsProver = ParamsKZG<E>;
+    type ParamsVerifier = ParamsVerifierKZG<E>;
+
+    fn new_params(k: u32) -> Self::ParamsProver {
+        ParamsKZG::new(k)
+    }
+
+    fn read_params<R: io::Read>(reader: &mut R) -> io::Result<Self::ParamsProver> {
+        ParamsKZG::read(reader)
+    }
+}
+
+// TODO: see the issue at https://github.com/appliedzkp/halo2/issues/45
+// So we probably need much smaller verifier key. However for new bases in g1 should be in verifier keys.
+/// KZG multi-open verification parameters
+pub type ParamsVerifierKZG<C> = ParamsKZG<C>;
+
+impl<'params, E: Engine + Debug> Params<'params, E::G1Affine, MSMKZG<E>> for ParamsKZG<E> {
+    fn k(&self) -> u32 {
+        self.k
+    }
+
+    fn n(&self) -> u64 {
+        self.n
+    }
```

Comment:
These can be const fns

Reply:
Why so?? They return `u64` no??? 
Which is the issue?

Reply:
Then we would need to add const generics to `Params` trait and change lots of things in signatures

Reply:
I dod not see that they were trait fns... My bad mate.